src/wasm/wasm-validator.cpp

```diff
@@ -1957,6 +1958,19 @@ void FunctionValidator::visitRefFunc(RefFunc* curr) {
   shouldBeTrue(!!func, curr, "function argument of ref.func must exist");
 }
 
+void FunctionValidator::visitRefEq(RefEq* curr) {
+  shouldBeTrue(
+    getModule()->features.hasGC(), curr, "ref.eq requires gc to be enabled");
+  shouldBeTrue(curr->left->type == Type::unreachable ||
+                 Type::isSubType(curr->left->type, Type::eqref),
+               curr->left,
+               "ref.eq's left argument should be a subtype of eqref");
```

Comment:
There is a helper function `shouldBeSubTypeOrFirstIsUnreachable` that you could use here.